internal/buffer/save.go

```diff
@@ -79,10 +80,34 @@ func overwriteFile(name string, enc encoding.Encoding, fn func(io.Writer) error,
 	if withSudo {
 		// wait for dd to finish and restart the screen if we used sudo
 		err := cmd.Wait()
+		if err != nil {
+			screen.TempStart(screenb)
+			return err
+		}
+
+		cmd = exec.Command(config.GlobalSettings["sucmd"].(string), "mv", tmpName, name)
+
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+		go func() {
+			<-c
+			cmd.Process.Kill()
+		}()
+
+		if e := cmd.Start(); e != nil && err == nil {
+			screen.TempStart(screenb)
+			return err
+		}
+
+		err = cmd.Wait()
 		screen.TempStart(screenb)
 		if err != nil {
 			return err
 		}
+	} else {
+		if err == nil {
+			err = os.Rename(tmpName, name)
```

Comment:
There are legitimate cases when rename fails (or some previous steps fail, e.g. creating the `.tmp` file) but it is actually perfectly possible to save the file, so we should still try our best to save it, rather than just return an error to the user. For example:

* Micro has no write permissions to the directory where the file is, so it cannot create other files in this directory and cannot change the name of the file. But it has write permissions to the file itself, so it can overwrite the file contents.
* The file itself is a mountpoint (e.g. a bind mountpoint), so technically it is on a different filesystem than other files in the same directory, so rename fails.

So, what to do? If the buffer has been successfully written to the temporary file and then `os.Rename()` failed, it means we already have a backup copy, so we can now overwrite the user's file as well. If overwrite successfully completes (including fsync!), we can remove the temporary file. If overwrite fails, we should keep the temporary file. And we should probably rename it to the backup file in `~/.config/micro/backups` then? (And if this rename, in turn, fails (e.g. if `~/.config/micro/backups` is on a different filesystem), we should try to copy & remove then? hmmm...)

But if it was not `os.Rename()` but some previous step that failed, e.g. if we failed to create `.tmp` file due to lack of permission, what then? Then we should probably be able to create a temporary file in `~/.config/micro/backups` instead. Actually, such a temporary file should be already there, - it is the backup file.

All this makes me think: as you noted yourself in your TODO, we should try to reuse the backup file if possible; and given the above, it seems we'd better *just* use the backup, no other temporary file. So the algorithm can be simple and clean:

1. If the `backup` option is on, synchronize the backup file with the buffer.
2. If the `backup` option is off (no backup file yet), create the backup file.
3. Rename the backup file to the user's file.
4. If rename failed, overwrite the user's file with the buffer.
5. If overwrite failed, don't remove the backup.

The `.tmp` approach might have some advantages over the above simplified approach (the reduced likelihood of rename failure, e.g. if the user works a lot file files on a different filesystem than `~/.config/micro`), but they seem to be outweighed by the increased likelihood of rename failure in other cases (e.g. lack of permission to the directory), the increased complexity and mess, and the fact that we may leave nasty `.tmp` files around in the user's directories.

Reply:
BTW looks like vim does more or less the same: https://github.com/vim/vim/blob/master/src/bufwrite.c#L1179

Moreover, if the file is a symlink, it doesn't seem to try to resolve it, it immediately falls backs to overwriting the file from the backup instead of renaming. I guess we can do the same, for simplicity (after all, a symlink is just a corner case, and we need to support the overwrite fallback anyway).

OTOH, I guess it doesn't hurt if we do resolve the symlink. E.g. we can move that to a separate `resolveSymlink()` function, so the code will not become much more complicated.

P.S. And nano, for that matter, doesn't seem to use rename at all: https://git.savannah.gnu.org/cgit/nano.git/tree/src/files.c#n1748
It just always overwrites the file from the temporary file, even though it is slower and less convenient.

Reply:
P.S. I hope we don't need to do all those scrupulous proactive checks like `st.st_dev != st_old.st_dev` that vim does, we can just try `os.Rename()` and fall back to overwrite if it fails. With the exception for symlinks, which we need to check or resolve beforehand, to prevent replacing them with regular files.

BTW... what about FIFOs, sockets, device files, ...?

Reply:
> BTW... what about FIFOs, sockets, device files, ...?

`micro` seems to have serious trouble with such files anyway, it crashed with a fifo, it hung with a fd...
`vim` "opens" them and tells, that these are no files.

Reply:
> If the `backup` option is off (no backup file yet), create the backup file.

Since the `backupThread` is async (with possible delay) we can create the backup anyway and don't need to consider the option. For the removal of the backup the option is important then, right?

Reply:
> Since the `backupThread` is async (with possible delay) we can create the backup anyway and don't need to consider the option. For the removal of the backup the option is important then, right?

Right, probably. That's implementation details.

Reply:
> Edit: I suggest to handle files only in case they're supported by `micro`. Means, we should prevent loading (and storing) files like `ModeDevice` + `ModeCharDevice`, `ModeIrregular`, etc.pp. because a crash or stuck is unacceptable.

Yep, and that should be easy: micro already refuses to open directories, so it can also just refuse to open those irregular files.

I mentioned those irregular files here just because we should ensure, in particular, that we don't replace them with regular files by rename when saving. But if micro never opens those files in the first place, that trouble would be already solved... Hmm, wait, no: the file may change from a regular file to an irregular one in the meantime. So we still need to check that at the moment of save as well?

Reply:
Our goal is only to ensure that the original file is never left in a damaged state (or in the worst case, it is left in a damaged state, but we have a backup so we can restore it), right? If `/home` is full so we cannot do anything, it is not our problem.

Reply:
Ok, that's a position. But even then we can only proceed in the moment the backup was successful, otherwise we would take a broken backup as source to overwrite a healthy target.

> Our goal is only to ensure that the original file is never left in a damaged state (or in the worst case, it is left in a damaged state, but we have a backup so we can restore it), right?

Do we then really have? In the moment we ignore running `/home` out of space and accept damaging the backup we've no "plan b" in the moment something goes wrong with the target.

Reply:
We should make sure to avoid this situation. If the backup was unsuccessful, we should *not* overwrite the target file with it, we should just tell the user that we cannot save the file. Also I think we should remove the broken backup.

Reply:
Ok, I can live with that. It's the simpler approach and we don't need to use temporary files.